core/shared/src/main/scala/cats/parse/RadixNode.scala

```diff
@@ -22,85 +22,170 @@
 package cats.parse
 
 import cats.data.NonEmptyList
-
+import cats.kernel.Semilattice
 import scala.annotation.tailrec
 
-private[parse] class RadixNode(
-    val fsts: Array[Char],
+private[parse] final class RadixNode(
+    matched: String,
+    bitMask: Int,
     // the prefixes are the rest of the string after the fsts (not including the fsts Char)
-    val prefixes: Array[String],
-    val children: Array[RadixNode],
-    val word: Boolean
+    prefixes: Array[String],
+    children: Array[RadixNode]
 ) {
-  override def toString(): String =
-    s"RadixNode(${fsts.mkString("[", ", ", "]")}, ${children.mkString("[", ", ", "]")}, $word)"
-}
+  override def toString(): String = {
+    def list[A](ary: Array[A]): String = ary.mkString("[", ", ", "]")
+    val ps = list(prefixes)
+    val cs = list(children)
+    s"RadixNode($matched, $bitMask, $ps, $cs)"
+  }
 
-private[parse] object RadixNode {
+  /** @return
+    *   all strings that are in this RadixNode
+    */
+  def allStrings: List[String] = {
+    // matched may be null (meaning there is not yet a partial match)
+    // or it may be non-null: there is a partial match
+    // if it is non-null it may be duplicated in children
+    val rest = children.iterator.flatMap {
+      case null => Nil
+      case c => c.allStrings
+    }
+
+    if (matched eq null) rest.toList
+    else (matched :: rest.filterNot(_ == matched).toList)
+  }
+
+  /** If this matches, return the new offset, else return -1
+    *
+    * @param str
+    *   the string to match against this RadixNode
+    * @param offset
+    *   the initial offset
+    * @return
+    *   the new offset after a match, or -1
+    */
+  def matchAt(str: String, off: Int): Int =
+    matchAtOrNull(str, off) match {
+      case null => -1
+      case nonNull => nonNull.length
```

Comment:
Actually I think this is wrong. It has to be offset + NonNull.length, or the comment is wrong. Maybe we should just fix the comment, but also the Parser is assuming it returns the new offset not the length.

I'll send a PR with a failing test and then fix.

I think this is happening because we are only testing at offset 0.

Reply:
Nice catch!

Being `offset + nonNull.length` makes more sense and the consistency with `matchAtOrNull` reflects it (which needs a fix).

These make more sense now 
```scala
            assert(len <= targ.length, s"len = $len, off = $off")
            assertEquals(left, targ.substring(off, len), s"len = $len, left = $left")
```